File: packages/adapter-typescript/src/scanner.ts
/**
 * File discovery + parsing of the TS/JS sources via the TypeScript
 * compiler API.
 *
 * Parse-only: no type resolution, no tsconfig, no npm install needed in the
 * target project. `relPath` is project-relative with '/' separators and the
 * deterministic sort key everywhere.
 */

import { lstatSync, readFileSync, readdirSync, statSync } from 'node:fs';
import { join } from 'node:path';
import ts from 'typescript';
import type { AdapterConfig } from './config.js';
import { AdapterException, type SourceRef } from './graph-model.js';

/** Source extensions the adapter understands (TS and JS, each with JSX). */
const SOURCE_EXTENSIONS = ['.ts', '.tsx', '.mts', '.cts', '.js', '.jsx', '.mjs', '.cjs'];

/**
 * Directories that are never scanned — regardless of the include globs
 * (necessary here unlike in the Dart adapter because node_modules/build
 * output sit next to the sources in TS/JS projects). Dot directories (.git,
 * .next, …) are always skipped.
 */
const EXCLUDED_DIRS = new Set(['node_modules', 'dist', 'build', 'out', 'coverage']);

/** A scanned source file with its parsed SourceFile. */
export class ScannedFile {
  #lineStarts: number[] | undefined;

  constructor(
    readonly relPath: string,
    readonly content: string,
    readonly sourceFile: ts.SourceFile,
  ) {}

  /** 1-based line of a character offset. */
  lineOf(pos: number): number {
    return this.sourceFile.getLineAndCharacterOfPosition(pos).line + 1;
  }

  /**
   * Offset of the line start of a 1-based line — counts only '\n' (like the
   * line-based comment parser), NOT TypeScript's line map: that one also
   * counts U+2028/U+2029/lone '\r' as line breaks, which would shift the
   * mapping of @journey blocks.
   */
  offsetOfLine(line: number): number {
    if (this.#lineStarts === undefined) {
      const starts = [0];
      for (let i = 0; i < this.content.length; i++) {
        if (this.content[i] === '\n') starts.push(i + 1);
      }
      this.#lineStarts = starts;
    }
    return this.#lineStarts[line - 1] ?? this.content.length;
  }

  refAt(pos: number, symbol?: string): SourceRef {
    return {
      file: this.relPath,
      line: this.lineOf(pos),
      ...(symbol !== undefined ? { symbol } : {}),
    };
  }
}

/**
 * Glob → RegExp over the posix-relative path: `**` matches across '/'
 * boundaries, `*`/`?` within a segment (like package:glob in the
 * Dart adapter).
 */
export function globToRegExp(pattern: string): RegExp {
  let out = '';
  for (let i = 0; i < pattern.length; i++) {
    const ch = pattern[i]!;
    if (ch === '*') {
      if (pattern[i + 1] === '*') {
        out += '.*';
        i++;
      } else {
        out += '[^/]*';
      }
    } else if (ch === '?') {
      out += '[^/]';
    } else if ('\\^$.|+()[]{}'.includes(ch)) {
      out += `\\${ch}`;
    } else {
      out += ch;
    }
  }
  return new RegExp(`^${out}$`);
}

function scriptKindOf(relPath: string): ts.ScriptKind {
  if (relPath.endsWith('.tsx')) return ts.ScriptKind.TSX;
  if (relPath.endsWith('.ts') || relPath.endsWith('.mts') || relPath.endsWith('.cts')) {
    return ts.ScriptKind.TS;
  }
  // .js/.jsx/.mjs/.cjs: parse as JSX — JSX in .js is common in the React
  // ecosystem, and plain JS parses unchanged under ScriptKind.JSX.
  return ts.ScriptKind.JSX;
}

/** Recursive file list as posix-relative paths, deterministically sorted. */
function listFiles(root: string): string[] {
  const result: string[] = [];
  const walk = (dir: string, relPrefix: string): void => {
    let entries: string[];
    try {
      entries = readdirSync(dir);
    } catch {
      return;
    }
    for (const entry of entries) {
      const rel = relPrefix === '' ? entry : `${relPrefix}/${entry}`;
      const abs = join(dir, entry);
      let stat;
      try {
        // lstat: symlinks are not followed (as in the Dart scanner).
        stat = lstatSync(abs, { throwIfNoEntry: false });
        if (stat === undefined) continue;
      } catch {
        continue;
      }
      if (stat.isDirectory()) {
        if (entry.startsWith('.') || EXCLUDED_DIRS.has(entry)) continue;
        walk(abs, rel);
      } else if (stat.isFile()) {
        result.push(rel);
      }
    }
  };
  walk(root, '');
  result.sort();
  return result;
}

/**
 * Collects all source files under the include globs and parses them.
 * Syntax errors do not abort: the TypeScript compiler API parses fault
 * tolerantly, the analysis is best effort.
 */
export function scanProject(
  projectDir: string,
  config: AdapterConfig,
  warn: (message: string) => void,
): ScannedFile[] {
  let rootStat;
  try {
    rootStat = statSync(projectDir, { throwIfNoEntry: false });
  } catch {
    rootStat = undefined;
  }
  if (rootStat === undefined || !rootStat.isDirectory()) {
    throw new AdapterException([`Project directory not found: ${projectDir}`]);
  }

  const includes = config.include.map(globToRegExp);
  const files: ScannedFile[] = [];
  for (const rel of listFiles(projectDir)) {
    if (!SOURCE_EXTENSIONS.some((ext) => rel.endsWith(ext))) continue;
    if (!includes.some((re) => re.test(rel))) continue;

    let content: string;
    try {
      content = readFileSync(join(projectDir, ...rel.split('/')), 'utf8');
    } catch {
      warn(`Warning: ${rel} is not readable — skipped.`);
      continue;
    }
    const sourceFile = ts.createSourceFile(
      rel,
      content,
      ts.ScriptTarget.Latest,
      /* setParentNodes */ true,
      scriptKindOf(rel),
    );
    // parseDiagnostics is not a public API, but it is the only way to get at
    // the parse errors without a Program/TypeChecker — access defensively.
    const parseDiagnostics = (sourceFile as unknown as { parseDiagnostics?: unknown[] })
      .parseDiagnostics;
    if (Array.isArray(parseDiagnostics) && parseDiagnostics.length > 0) {
      warn(`Warning: ${rel} contains syntax errors; analysis is best effort.`);
    }
    files.push(new ScannedFile(rel, content, sourceFile));
  }
  return files;
}

File: dart/ductus/lib/src/adapter/scanner.dart
/// Collects and parses the target project's .dart files (parse-only, without
/// resolution — so the target project needs no `pub get`).
library;

import 'dart:io';

import 'package:analyzer/dart/analysis/utilities.dart';
import 'package:analyzer/dart/ast/ast.dart';
import 'package:analyzer/source/line_info.dart';
import 'package:glob/glob.dart';
import 'package:path/path.dart' as p;

import 'graph_model.dart';
import 'config.dart';

/// A parsed source file. [relPath] is project-relative with '/' separators
/// and serves everywhere as the deterministic sort key.
class ScannedFile {
  final String relPath;
  final String content;
  final CompilationUnit unit;
  final LineInfo lineInfo;

  const ScannedFile({
    required this.relPath,
    required this.content,
    required this.unit,
    required this.lineInfo,
  });

  int lineOf(int offset) => lineInfo.getLocation(offset).lineNumber;

  SourceRef refAt(int offset, {String? symbol}) =>
      SourceRef(file: relPath, line: lineOf(offset), symbol: symbol);
}

/// Returns all .dart files under the include patterns, deterministically
/// sorted by relative path and analyzed parse-only.
List<ScannedFile> scanProject(
  String projectDir,
  AdapterConfig config,
  void Function(String) warn,
) {
  final root = Directory(projectDir);
  if (!root.existsSync()) {
    throw AdapterException(['Project directory not found: $projectDir']);
  }
  final globs = config.include
      .map((pattern) => Glob(pattern, context: p.posix))
      .toList();

  final relPaths = <String>[];
  for (final entity in root.listSync(recursive: true, followLinks: false)) {
    if (entity is! File || !entity.path.endsWith('.dart')) continue;
    final rel = p.posix.joinAll(p.split(p.relative(entity.path, from: root.path)));
    if (globs.any((g) => g.matches(rel))) relPaths.add(rel);
  }
  relPaths.sort();

  final files = <ScannedFile>[];
  for (final rel in relPaths) {
    final abs = p.join(root.path, p.joinAll(p.posix.split(rel)));
    final content = File(abs).readAsStringSync();
    final result = parseString(
      content: content,
      path: abs,
      throwIfDiagnostics: false,
    );
    if (result.errors.isNotEmpty) {
      // Best effort: use a file with syntax errors anyway, but report it.
      warn('Warning: $rel contains syntax errors; analysis is best effort.');
    }
    files.add(ScannedFile(
      relPath: rel,
      content: content,
      unit: result.unit,
      lineInfo: result.lineInfo,
    ));
  }
  return files;
}
